### ISSUE TITLE
Make IconLayer compatible with binary MVTLayer

### ISSUE THREAD

**Issue by felixpalmer** [visgl/deck.gl#5815]
#### Target Use case
The `binary` version of the `MVTLayer` (default as of v8.5) provides a faster way to load data, but only supports `GeoJsonLayer` in the renderSubLayers callback.

#### Proposed feature
`IconLayer` is made compatible with the `binary` `MVTLayer`

#### To Do List 
  - [ ] Add label and assign to milestone 
  - [ ] Coding
  - [ ] Doc update
  - [ ] What’s new update 
  - [ ] Test

**Comment by felixpalmer** [comment]
I've managed to get the position data passed through with a fairly simple modification of `updateState` in `IconLayer`:

```javascript
updateState({oldProps, props, changeFlags}) {
  const {data} = props;
  const binary = data && 'points' in data && 'polygons' in data && 'lines' in data;

  if (binary) {
    const layerProps = createLayerPropsFromBinary(props.data, this.encodePickingColor);
    this.props.data = layerProps.points.data;
  } else {
    super.updateState({props, oldProps, changeFlags});
  }
  ...
}
```

However providing support for accessors like `getSize` is more tricky. It is not possible to simply modify the accessor props like so, as they are read-only

```javascript
updateState({oldProps, props, changeFlags}) {
  ...
  if (binary) {
    const layerProps = createLayerPropsFromBinary(props.data, this.encodePickingColor);
    this.props.data = layerProps.points.data;
    
    this.props.getSize = this._getBinaryAccessor(this.props.getSize);
  } else {
 ...
}
    
...

_getBinaryAccessor(accessor) {
  if (typeof accessor !== 'function') {
    return accessor;
  }

  return (object, info) => {
    const {data, index} = info;
    const feature = binaryToFeatureForAccesor(data, index);
    return accessor(feature, info);
  };
}
```

**Comment by felixpalmer** [comment]
Any input on how to best achieve this @Pessimistress? I've verified by temporarily unfreezing the props that the data and properties end up in the correct place and the layer is drawn, but naturally this is not a solution. I'm not sure design-wise where to best place the `binaryToFeatureForAccesor` conversion function so to the consumer of the library the fact that the binary representation is being used is transparent (like we have in the `GeoJsonLayer`)

**Comment by Pessimistress** [comment]
You are not supposed to modify props once a layer is created. You should manipulate props before creating the layer in `renderSubLayers`. The sub layer class used for points in `GeoJsonLayer` can be replaced with `_subLayerProps.points`.

Also `IconLayer` should not have special treatment for GeoJSON data, binary or not. You need to use `data.attributes` to pass binary data, just like what we are doing with the `ScatterplotLayer`.

**Comment by felixpalmer** [comment]
OK so if `IconLayer` shouldn't have special treatment, the binary data needs to be set to `props.data.attributes` in the `MVTLayer`, to be passed down to the sublayer. I can do this by modifying props in the `renderSubLayers` function : https://github.com/visgl/deck.gl/blob/7862bb982f495ac9e2ae41075d8e83f8fdf3d24d/modules/geo-layers/src/mvt-layer/mvt-layer.js#L137

like so:

```javascript
renderSubLayers(props) {
  ...
  // Place binary data into attributes
  const {data} = props;
  const binary = data && 'points' in data && 'polygons' in data && 'lines' in data;
  if (binary) {
    const layerProps = createLayerPropsFromBinary(props.data, this.encodePickingColor);
    props.data = layerProps.points.data;
    // props.data.attributes is now {getPosition: {…}, pickingColors: {…}}
  }
   
  return super.renderSubLayers(props);
}
```

And it works for `IconLayer`, but it breaks the `GeoJSONLayer` because this is expecting the full props prior to transformation in order to extract all the other layer props (for lines and polygons).

Is there some way I can know ahead of time what sublayer we will be rendering? Otherwise it feels like I'm a bit stuck here. Or some other way to make this work with both types of sublayers?

**Comment by felixpalmer** [comment]
It is possible to do this:

```javascript
renderSubLayers(props) {
  ...
  let subLayers = super.renderSubLayers(props);

  if (!(subLayers instanceof GeoJsonLayer)) {
    const {data} = props;
    const binary = data && 'points' in data && 'polygons' in data && 'lines' in data;
    if (binary) {
      const layerProps = createLayerPropsFromBinary(props.data, this.encodePickingColor);
      props.data = layerProps.points.data;
    }
    subLayers = super.renderSubLayers(props);
  }

  return subLayers;
}

But I'm not super happy about it from a design perspective.
```

**Comment by Pessimistress** [comment]
https://deck.gl/docs/api-reference/core/composite-layer#_sublayerprops

```js
new MVTLayer({
  renderSubLayer: props => new GeoJsonLayer({
    ...props,
    _subLayerProps: {
      points: {
        type: IconLayer,
        ...
      }
  }
});
```

**Comment by felixpalmer** [comment]
Thanks for the pointer, I didn't quite understand what you meant before. Even with this approach all doesn't work as expected, namely the accessor functions aren't getting the properties passed through, specifically:

```javascript
getSize: (d) => {
  // d.properties is undefined, d is {geometry: {…}, __source: {…}}
}
```
Whereas if the `IconLayer` is used directly the properties are passed through. Arguably this is a bug that could be fixed.

However, in addition I feel the API via `_subLayerProps` is not so intuitive. Directly supplying an `IconLayer` to `renderSubLayers` currently works (see example below) when `binary` is false. Do you not think it is desirable from the point of view of the user of the library to support this for the `binary` option also, rather than forcing them to use `MVTLayer` + `_subLayerProps`?

```javascript
const layer3 = new MVTLayer({
  id: 'layer3',
  binary: false,
  data: values[1].urls,
  getPosition: f => f.geometry.coordinates,
  renderSubLayers: props => {
    return new IconLayer(props, {
      ...
      getSize: (d, info) => {
        return 10 + d.properties.cartodb_id;
      }
    });
  }
});
```

**Comment by Pessimistress** [comment]
Correct, the existing solution still requires a lot of plumbing. One option is to add a `pointType: 'circle' | 'icon'` prop to the `GeoJsonLayer`, and handle the attribute/accessor mapping there.

**Comment by alasarr** [comment]
>Correct, the existing solution still requires a lot of plumbing. One option is to add a pointType: 'circle' | 'icon' | 'text' prop to the GeoJsonLayer, and handle the attribute/accessor mapping there.

I think we can go with this option

**Comment by alasarr** [comment]
How about if we launch an exception if someone uses MVT layer with binary true and `renderSubLayers` not instance of `GeoJsonLayer`? It's explained in the doc, but the exception will make this clear.

**Comment by felixpalmer** [comment]
I'm also in favour of using an explicit `pointType` attribute as all other options so far seem brittle and hard to discover for the user. This way it is explicit. I'll put together a PR

**Comment by felixpalmer** [comment]
@Pessimistress @alasarr could you take a look at https://github.com/visgl/deck.gl/pull/5835 to see if you agree with this approach?

**Comment by Pessimistress** [comment]
Now available in 8.5.0-alpha.10.
